### PR TITLE
Allow extension cluster menus to have a parent

### LINKED ISSUE #1442: Possibility to register cluster sub-menus
Currently extension-api allows only to register top-level menu items to cluster menu. Extension-api should allow only registration of sub-items.

### PR DESCRIPTION
**Pull request by jakolehm**
- introduces `id` and `parentId` properties (both optional)
- TabLayout is created automatically if cluster menu item has sub-items

Example:

```ts
    clusterPageMenus = [
    {
      id: "example",
      title: "Example Extension",
      components: {
        Icon: ExampleIcon,
      }
    },
    {
      parentId: "example",
      target: { pageId: "example1", params: {} },
      title: "Sub menu 1",
      components: {
        Icon: ExampleIcon,
      }
    },
    {
      parentId: "example",
      target: { pageId: "example2", params: {} },
      title: "Sub menu 2",
      components: {
        Icon: ExampleIcon,
      }
    }
  ]
```

produces:

image

Fixes #1442

### CODE REVIEW

File: src/renderer/components/app.tsx [hunk 1]
```diff
@@ -72,9 +75,58 @@ export class App extends React.Component {
...
+    });
+    if (routes.length > 0) {
+      routes.unshift({
+        routePath: page.routePath,
+        url: getExtensionPageUrl({ extensionId: page.extensionId, pageId: page.id, params: menuItem.target.params }),
+        title: <Trans>Overview</Trans>,
+        component: page.components.Page,
+        exact: page.exact
+      })
+    }
```
Comment (Nokel81) [inline src/renderer/components/app.tsx]: Why do we force an `Overview` page, that seems weird.

Comment (jakolehm) [reply src/renderer/components/app.tsx]: See the attached screenshot. Extension basically does not know if something registers as a sub-menu... it's better to create consistent UI with "Overview" than duplicate top-level menu name in a first sub-menu item.

Comment (Nokel81) [reply src/renderer/components/app.tsx]: Wouldn't the API be better then if it does know if it registers as a sub menu? I see two main problems: one, is that there is no way to change that title text; two, there doesn't seem to be a way to remove this, if the extension developer doesn't want it.

Comment (jakolehm) [reply src/renderer/components/app.tsx]: Example:
- extension A registers a menu "Foo"
- extension B registers a sub-menu "Bar" to "Foo"

Comment (Nokel81) [reply src/renderer/components/app.tsx]: Okay, so then have a tab layout of size 1.

Comment (Nokel81) [reply src/renderer/components/app.tsx]: Also, I don't see any handling of the following case:

- extension A registers a menu "Foo"
- extension B registers a sub-menu "Bar" to "Fooo" (by accident). 

should an error be reported (somewhere)?

Comment (jakolehm) [reply src/renderer/components/app.tsx]: > Okay, so then have a tab layout of size 1.

Not sure if I understand your point. Our menu does not really work like that?

Comment (jim-docker) [reply src/renderer/components/app.tsx]: But other (builtin) submenus don't have an overview page

What about something like:
```
  clusterPageMenus = [
    {
      target: { submenu: "example" },
      title: "Example Extension",
      components: {
        Icon: ExampleIcon,
      }
    },
    {
      parentId: "example",
      target: { pageId: "example1", params: {} },
      title: "Overview",
      components: {
        Icon: ExampleIcon,
      }
    },
    {
      parentId: "example",
      target: { pageId: "example2", params: {} },
      title: "Sub menu 1",
      components: {
        Icon: ExampleIcon,
      }
    }
  ]
```

Comment (Nokel81) [reply src/renderer/components/app.tsx]: I agree with Jim, this shouldn't be "automagically" added. And with regards to the tab layout not working like that. It seems very strange that the tab layout should care how many entries it has at all.

Comment (jakolehm) [reply src/renderer/components/app.tsx]: Ok, based on discussions (on Slack) I'll limit scope of this PR to work only within an extension. We can improve (and introduce more possibilities) on follow up PRs.

Comment (Nokel81) [reply src/renderer/components/app.tsx]: Sounds good

**Comment by jakolehm** [comment]
@Nokel81 @jim-docker PTAL (example in description is updated to match current logic)

File: src/renderer/components/app.tsx [hunk 2]
```diff
@@ -72,9 +75,48 @@ export class App extends React.Component {
     return workloadsURL();
   }
 
+  getTabLayoutRoutes(menuItem: ClusterPageMenuRegistration) {
+    const routes: TabLayoutRoute[] = [];
+    if (!menuItem.id) {
+      return routes;
+    }
+    clusterPageMenuRegistry.getSubItems(menuItem).forEach((item) => {
+      const page = clusterPageRegistry.getByPageMenuTarget(item.target);
```
Comment (Nokel81) [inline src/renderer/components/app.tsx:84]: There doesn't seem to be an restrictions here about registering an item to a different extension.

File: src/renderer/components/layout/sidebar.tsx [hunk 3]
```diff
@@ -76,6 +76,29 @@ export class Sidebar extends React.Component<Props> {
...
+      return [];
+    }
+    const routes: TabLayoutRoute[] = [];
+
+    clusterPageMenuRegistry.getSubItems(menu).forEach((subItem) => {
+      const subPage = clusterPageRegistry.getByPageMenuTarget(subItem.target);
+      console.log(subItem, subPage)
+      if (subPage) {
+        routes.push({
+          routePath: subPage.id,
```
Comment (jim-docker) [inline src/renderer/components/layout/sidebar.tsx]: should this be `subPage.routePath`?

File: src/renderer/components/layout/sidebar.tsx [hunk 4]
```diff
@@ -76,6 +76,29 @@ export class Sidebar extends React.Component<Props> {
     });
   }
 
+  getTabLayoutRoutes(menu: ClusterPageMenuRegistration, page: RegisteredPage): TabLayoutRoute[] {
```
Comment (jim-docker) [inline src/renderer/components/layout/sidebar.tsx]: `page` is not used

File: src/renderer/components/app.tsx (hunk 2 above)
Comment (jakolehm) [reply src/renderer/components/app.tsx:84]: Fixed.

File: src/renderer/components/layout/sidebar.tsx (hunk 3 above)
Comment (jakolehm) [reply src/renderer/components/layout/sidebar.tsx]: Yes, good catch!

File: src/renderer/components/layout/sidebar.tsx (hunk 4 above)
Comment (jakolehm) [reply src/renderer/components/layout/sidebar.tsx]: Fixed.

File: src/renderer/components/layout/sidebar.tsx [hunk 5]
```diff
@@ -76,6 +76,29 @@ export class Sidebar extends React.Component<Props> {
...
+  getTabLayoutRoutes(menu: ClusterPageMenuRegistration): TabLayoutRoute[] {
+    console.log("Menu", menu)
+    if (!menu.id) {
+      return [];
+    }
+    const routes: TabLayoutRoute[] = [];
+
+    clusterPageMenuRegistry.getSubItems(menu).forEach((subItem) => {
+      const subPage = clusterPageRegistry.getByPageMenuTarget(subItem.target);
+      console.log(subItem, subPage)
```
Comment (Nokel81) [inline src/renderer/components/layout/sidebar.tsx]: please remove
```suggestion
```

File: src/renderer/components/layout/sidebar.tsx [hunk 6]
```diff
@@ -76,6 +76,29 @@ export class Sidebar extends React.Component<Props> {
     });
   }
 
+  getTabLayoutRoutes(menu: ClusterPageMenuRegistration): TabLayoutRoute[] {
+    console.log("Menu", menu)
```
Comment (Nokel81) [inline src/renderer/components/layout/sidebar.tsx]:
```suggestion
```

File: src/renderer/components/layout/sidebar.tsx [hunk 7]
```diff
@@ -191,17 +212,25 @@ export class Sidebar extends React.Component<Props> {
...
               return (
                 <SidebarNavItem
                   key={pageUrl} url={pageUrl}
-                  text={title} icon={<Icon/>}
+                  text={menuItem.title} icon={<menuItem.components.Icon/>}
                   isActive={isActive}
+                  subMenus={tabRoutes}
                 />
               );
             })}
```
Comment (Nokel81) [inline src/renderer/components/layout/sidebar.tsx]: Last thing, probably minor. But this is getting rather long so (imo) it should be in its own render function. Maybe `renderRegisteredMenus`?

File: src/renderer/components/layout/sidebar.tsx [hunk 8]
```diff
@@ -191,17 +212,25 @@ export class Sidebar extends React.Component<Props> {
             >
               {this.renderCustomResources()}
```
Comment (Nokel81) [inline src/renderer/components/layout/sidebar.tsx:238]: I just noticed that there is no place here that renders sub menus registered to Lens native menus. Is that still on the table for this PR?

Comment (jakolehm) [reply src/renderer/components/layout/sidebar.tsx:238]: Out-of-scope, it needs more refactoring of Lens internals.

Comment (Nokel81) [reply src/renderer/components/layout/sidebar.tsx:238]: Makes sense

File: src/renderer/components/layout/sidebar.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/components/layout/sidebar.tsx]: If not, I will approve

Comment (jakolehm) [reply src/renderer/components/layout/sidebar.tsx]: Fixed.

**Review by Nokel81 — APPROVED** [review]

**Pull request merged**